core/vm/state_manager.go

```diff
@@ -0,0 +1,113 @@
+package vm
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+type stateManagerFunction func(*EVM, *Contract, []byte) ([]byte, error)
+type methodId [4]byte
+
+var funcs = map[string]stateManagerFunction{
+	"getStorage(address,bytes32)":            getStorage,
+	"setStorage(address,bytes32,bytes32)":    setStorage,
+	"getOvmContractNonce(address)":           getOvmContractNonce,
+	"getCodeContractBytecode(address)":       getCodeContractBytecode,
+	"getCodeContractHash(address)":           getCodeContractHash,
+	"getCodeContractAddress(address)":        getCodeContractAddress,
+	"associateCodeContract(address,address)": associateCodeContract,
+	"incrementOvmContractNonce(address)":     incrementOvmContractNonce,
+}
+var methodIds map[[4]byte]stateManagerFunction
+var executionMangerBytecode []byte
+
+func init() {
+	methodIds = make(map[[4]byte]stateManagerFunction, len(funcs))
+	for methodSignature, f := range funcs {
+		methodIds[MethodSignatureToMethodId(methodSignature)] = f
+	}
+}
+
+func MethodSignatureToMethodId(methodSignature string) [4]byte {
+	var methodId [4]byte
+	copy(methodId[:], crypto.Keccak256([]byte(methodSignature)))
+	return methodId
+}
+
+func callStateManager(input []byte, evm *EVM, contract *Contract) (ret []byte, err error) {
+	var methodId [4]byte
+	if len(input) == 0 {
+		return nil, nil
+	}
+	copy(methodId[:], input[:4])
+	ret, err = methodIds[methodId](evm, contract, input)
+	return ret, err
+}
+
+func setStorage(evm *EVM, contract *Contract, input []byte) (ret []byte, err error) {
+	address := common.BytesToAddress(input[4:36])
+	key := common.BytesToHash(input[36:68])
+	val := common.BytesToHash(input[68:100])
```

Comment:
Can we make assumptions about the size of `input`, or should we check to see that it is >= 100 in length?

Reply:
Ah yeah if I disable calls to the StateManager from anyone other than the EM (which I need to do), then we should be good wrt being able to make strong assumptions about input size

Reply:
I have added this! So we should be safe and not need to sanitize inputs

Reply:
It might be useful to add a comment above this function noting the assumption that this function can only be called by the execution manager so that our future selves don't forget.